text/3808-register-tool.md

```diff
@@ -0,0 +1,257 @@
+[`eslint`]: https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1
+[`review`]: https://github.com/Bogdanp/racket-review#usage
+[Roslyn analyzers]: https://johnnyreilly.com/eslint-your-csharp-in-vs-code-with-roslyn-analyzers#deactivate-linting-partially
+[gcc]: https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html
+[vendor attributes]:  https://en.cppreference.com/w/cpp/language/attributes
+[c-sharp attrs]: https://learn.microsoft.com/en-us/dotnet/standard/attributes/applying-attributes
+[active]: https://rustc-dev-guide.rust-lang.org/attributes.html#non-builtinactive-attributes
+[`Resyntax`]: https://docs.racket-lang.org/resyntax/Refactoring_Rules_and_Suites.html#(part._.Exercising_.Fine_.Control_.Over_.Comments)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How does this interact with [proc-macro lints][`proc_macro_lint`]?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- Proc macros wish to register custom lints; see [`proc_macro_lint`]. We would have to establish some mechanism to prevent overlapping namespaces. Perhaps `warn(::project::lint_name)` could refer to the proc macro and `warn(project::lint_name)` would refer to any registered tool (only when a `project` tool is regisetered; in the common case where no tool is registered, `project::` would still refer to the proc macro).
+- Projects may wish to have both a proc-macro crate with lints and a CLI with lints. To allow this, we would require `proc_macro_lint` to create an exhaustive list of lints that can be created, such that we can still run `unknown_lints` and do not need to create a new cooperation mechanism between `proc_macro_lint` and `register_lint_tool`, nor to require users of the project to distinguish the two with `::project` (see immediately above). We might still run into difficulty if the proc-macro lint namespace is only active while the proc-macro is expanding; it depends on how `proc_macro_lint` is specified. But I think it's ok to delay that discussion until `proc_macro_lint` gets an RFC.
+- We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.
+- Once [expression attributes] are stabilized, this would also allow tool attributes on expressions.
+- We could support registering tools through `Cargo.toml`, which Cargo would then pass to `rustc` using `--crate-attr` (similar to `--check-cfg`). We would have to consider how this interacts with duplicate registration; we don't want to error if a tool is configured both through Cargo.toml and in source code.
```

Comment:
Covering every crate in a package can be annoying otherwise which may be worth calling out.

Reply:
I think I'm going to remove the hard error on duplicate registration and only error if people try to register a built-in tool, which resolves this nicely. I'll also specify a mechanism for Cargo to pass this to rustc. how does `[package] tools = ["c2rust", "bevy_lint"]` sound?

Reply:
Creating a new top-level table is going to need a lot more elaboration, especially if its as generic as `tools`.

Should this RFC be split into two parts?
- rustc side
- cargo side